cloudflare/resource_cloudflare_zone_dnssec.go

```diff
@@ -41,7 +42,7 @@ func resourceCloudflareZoneDNSSECCreate(d *schema.ResourceData, meta interface{}
 	if err != nil {
 		return fmt.Errorf("error finding Zone DNSSEC %q: %s", zoneID, err)
 	}
-	if currentDNSSEC.Status != DNSSECStatusActive {
+	if currentDNSSEC.Status != DNSSECStatusActive && currentDNSSEC.Status != DNSSECStatusPending {
```

Comment:
it seems like we should invert this check and just check against `DNSSECStatusDisabled`, no?

Reply:
That was my first thought, but then I checked the API docs (https://api.cloudflare.com/#dnssec-properties) and there are two extra possible values for this field: `pending-disabled` and `error`. If I'm not wrong, neither of them represent a enabled status.

Reply:
that's fair - my assumption was we only wanted to attempt to update this value if it was explicitly disabled however, perhaps we also want it on `error` too. we don't want to attempt the `Create` if there is an operation in progress (pending, pending-disabled) or already active. in which case, i think we also need to extend this to include the `pending-disabled` status too.

could i also ask that you add these two other conditions to the constants so we have them all?

Reply:
I've been trying some tests. I don't know how to trigger an `error` state, so I'm not sure what it does mean or if we should try or not to create the resource.

`pending-disable` is an intermediate state before showing DNSSEC as fully disabled; the UI shows the following message:

> Your DNSSEC setup will be disabled as soon as we detect that the DS record has been removed from your registrar.

This state is effectively disabled and DNSSEC can be enabled again with no issues. I don't think we should skip `Create` in this state.

I can add two new constants, `DNSSECStatusPendingDisabled` and `DNSSECStatusError` but I don't think we will be using them (at least the first one, not sure about the `Error` one).

Reply:
I'm struggling to come up with scenarios where we wouldn't want to prevent attempting to recreate when the status is pending disabled but a handful o log scenarios where you wouldn't want successive calls to attempt to create the resource while it is pending cleanup. The only things I can think of are if you disabled it by accident?

Reply:
As I see it, `pending-disable` and `disable` are the same status, with the existence of the DS record (or rather, Cloudflare not having already detected its absence) on the domain as the only difference, but this is something "out of scope" of Cloudflare. For example, if we don't allow enabling DNSSEC in this state, `terraform taint cloudflare_zone_dnssec.example && terraform apply` would have different behaviour depending on if the DNS changes in the domain have propagated or not.

I always tend to assume that Terraform providers should behave exactly like direct API requests. If I can re-enable DNSSEC through the web UI or the API even in a `pending-disable`, why would Terraform block me to do it? It seems a little counter-intuitive.

Of course, almost 100% of the use cases are either accidents or testing, in normal execution I also don't see why anyone would do it. For example, my current use case is that we are developing our DNS related IaC code and at the same time, we are migrating our DNS to Cloudflare. I would not expect to wait until DS changes from our registrar are propagated everytime a `terraform destroy` is executed after a test so that the status is detected as `disabled` instead of `pending-disabled`.

If we don't allow enabling DNSSEC when in `pending-disabled` status, the same would apply to disabling it when in `pending` state. And this would be quite odd: do I need to add a DS record to a domain and wait for it to propagate before Terraform allows me to disable DNSSEC?

Reply:
I agree with you on the API compatibility. The difference though is that Terraform has lifecycle management which means in some places, it has to differ. I'm not strongly opposites enough here though so let's try it and see what happens.